### DISCUSSION TITLE
Show-InstallationWelcome -CloseApps - Separate different PS processes

### DISCUSSION THREAD

**Discussion by stephannn** [PSAppDeployToolkit/PSAppDeployToolkit#720] [Q&A] [answered]
Hi,

I am having a PowerShell WPF application that I would like to advise the user to close before an update. However, the function Show-InstallationWelcome -CloseApps only supports process names and no IDs. I can get my correct process with:

```
	$appRunningPowerShellProcess = "My process"
	if($appRunningPowerShellProcess){
		#$processRunning = $null

		foreach ($psp in $appRunningPowerShellProcess.Split(",")){
			Get-Process | Where-Object {($_.ProcessName -like "PowerShell*" -or $_.ProcessName -eq "pwsh") -and $_.MainWindowTitle -eq $psp } | ForEach-Object {$processRunning += $_}
		}
	}
```

but when I run the InstallationWelcome it will kill the complete PowerShell tree. One way would be creating an own exe so that I can check this process. Another would be the possibility to pass the process ID, but this isn's supporzed?!?! 

Has anyone experienced something like that or know how to deal with it?

Thanks
Stephan

**Comment by That-Annoying-Guy** [comment]
FYI: there is a better place for help at: https://discourse.psappdeploytoolkit.com/c/thetoolkit/general-discussion/11

That said, you are in luck. A couple weeks back I had the same issue.
I too tried to use the Window Title but that requires the code to see it to run in the same security context as the script you are trying to kill.
Instead use the commandline of the process like this:

```
[string]$process = "powershell.exe"
$AllPSprocesses = Get-WmiObject Win32_Process -Filter "name = '$process'" | Select-Object CommandLine,ProcessId
Write-Log $($AllPSprocesses | Out-String)
ForEach ($PSprocesses in $AllPSprocesses) {
	If ($PSprocesses.CommandLine -like "*IMITSelfHelp.ps1*") {
		Write-Log "killing [$($PSprocesses.CommandLine)]..."
		[Int]$ProcessID = $PSprocesses.ProcessId
		#WARNING: Stop-Process either doesn't work or blows up
		#$OutputText = Stop-Process -Id $ProcessID -Force -Debug -Confirm:$false -ErrorVariable ErrVar -OutVariable OutVar -WarningVariable WarnVar -ErrorAction SilentlyContinue | Out-String
		#Write-Log "[$OutputText] [$ErrVar] [$OutVar] [$WarnVar]"
		#Remove-variable OutputText,ErrVar,OutVar,WarnVar -ErrorAction SilentlyContinue
		Try {
			$ProcessToKill = Get-Process -Id $ProcessID
			$ProcessToKill.Kill()
		} catch {
			#In case killing the process fails, we want to know why.
			[String]$KillErrorMessage = @{'Message:' = "$($_.Exception.Message)";'ScriptStackTrace:'="$($_.ScriptStackTrace)";'InnerException:'="$($_.Exception.InnerException)"}
			Write-Log $KillErrorMessage
		}
	}
}
Start-Sleep -Seconds 3
```

**Comment by mjr4077au** [comment, marked as answer]
You'll be able to do filtration based on any Process property you wish in PSAppDeployToolkit 4.1.0 by way of https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/f0c8e576d492e70a51a32e871022a4aacd260aeb and https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/a63018e87f20d29aa6134c07b9505d3a79b38834. This should solve your original problem you were discussing.
